### ISSUE TITLE
Better docgen which includes exported symbols under ``when defined``

### ISSUE THREAD

**Issue by dom96** [nim-lang/Nim#2251]
The current `net` module's docs do not contain procedures for SSL because they are hidden away under `when defined(ssl)`, docgen should ideally detect this and mention it in the docs.

**Comment by Araq** [comment]
No, you should do 'when defined(ssl) or defined(docgen)'.

**Comment by dom96** [comment]
In what situation would you not want all exported symbols to be included in
the docs?

Specifying that extra piece of code is a PITA

**Comment by Araq** [comment]
So should the docgen simply assume every defined(symbol) is true then? That would be worse and weird:

```nim

when defined(debug): # disabled code
  proc doesntCompile*()  = error here
```

**Comment by dom96** [comment]
Why would you disable code like that?

The docgen shouldn't assume that. It should tell you "this procedure is only available when compiled with ``-d:whatever``"

**Comment by Araq** [comment]
I still think this proposal makes no sense whatsoever.

**Comment by dom96** [comment]
why do you think that?

**Comment by def-** [comment]
I agree with dom96. Many modules have optional functionality hidden behind defines. These should be in the documentation.

**Comment by Araq** [comment]
Because it's not clear at all what the docgen should do. You simply say "it should do what I have in mind", but this is not well defined at all.

**Comment by dom96** [comment]
That is not what I am saying. But if my suggestion is "not well defined" then let's define it more strongly.

The docgen should look at symbols defined under a ``when defined(x)`` where ``x`` can be any named identifier. If the symbols found under it are exported with an asterisk then the docgen should generate documentation for them **but** it should make it clear that these symbols are only available when compiled with ``--d:x``.

There are many options about *how* we will "make it clear that these symbols are only available when compiled with ``--d:x``". We don't need to define that yet.

**Comment by Araq** [comment]
> *The docgen should look at symbols defined under a when defined(x) where x can be any named identifier. If the symbols found under it are exported with an asterisk then the docgen should generate documentation for them ...*

This is actually quite impossible to do or requires a tremendous amount of effort.

**Comment by dom96** [comment]
Why is it so difficult?

**Comment by andreaferretti** [comment]
I think it is definitely simpler for library authors to

*  either use `or defined(docgen)`,
* or run `nim doc -d:whatever`

The latter can be easily automated into a nimble task `nimble docgen` (it is what I do for CUDA support in linalg). This way, authors have full control on the resulting documentation, and the compiler does not need to be more complex because of that

**Comment by mratsim** [comment]
I'm doing the same as andrea suggestion for Arraymancer for docgen.

Closely related issue but much harder to solve is that nimsuggest doesn't see anything behind `when defined(x):` so autocompletion of procs that depends on ssl, cuda, or any optional flag does not work and would require users of those libs to configure a custom nim.cfg.

**Comment by dom96** [comment]
Yeah, and IMO that's another reason that this should be supported. NimSuggest should be aware of these defines and let the user know that it is only available with ``-d:ssl`` or whatever.

**Comment by Araq** [comment]
Not gonna happen anytime soon.

**Comment by dom96** [comment]
That's not a reason to close this issue.

**Comment by Araq** [comment]
Yes it is, the issue tracker is not a list of "nice to have features" that nobody will ever work on.

**Comment by dom96** [comment]
I will work on this.
